feat: 32-see-the-route-list

Odonata971 opened the pull request:
- Moved client-related classes to a new entity package.
- Added route management features including Route, RoutePage, and RouteState.
- Implemented display logic for routes in the UI with a dedicated adapter.
- Enhanced Home activity to handle multiple fragments including routes.
- Created InfoRoute activity for detailed route information and actions.

Commit:
feat: restructure client and route management

- Moved client-related classes to a new entity package.
- Added route management features including Route, RoutePage, and RouteState.
- Implemented display logic for routes in the UI with a dedicated adapter.
- Enhanced Home activity to handle multiple fragments including routes.
- Created InfoRoute activity for detailed route information and actions.

Commit:
feat: Add client state management and UI updates

- Introduced `ClientState` enum to track visit status.
- Updated `Client` class with state handling methods.
- Enhanced timeline item display with client info and address.
- Modified route-related classes to utilize new state logic.
- Added drawable resources for visited and skipped states.
- Improved layout for timeline items, including dynamic dot colors.

Commit:
test: update parser tests for new data structure

- Adjusted test cases to reflect changes in JSON response format.
- Updated company names and coordinates in assertions.
- Changed expected client retrieval logic to match new structure.
- Modified route details to align with updated API response.

Commit:
feat: simplify timeline item display

- Removed unnecessary drawable creation for the last item.
- Added visibility handling for the timeline arrow.
- Cleaned up constructors in the timeline item class.
- Introduced a new XML drawable for the arrow design.

Commit:
feat: add route creation functionality

- Added a new activity for creating routes.
- Integrated route selection with itineraries.
- Updated the home screen to launch the new route activity.
- Enhanced existing classes to support route management and data handling.
- Included UI elements for user interaction in the layout.

Commit:
WIP on develop

WIP on feat/32-see-the-route-list

WIP on develop

WIP on develop

Commit:
feat: update client and itinerary handling

- Added new imports for Client and ClientArrayAdapter.
- Replaced hardcoded route key with a constant.
- Updated method signatures to use the new ItineraryArrayAdapter.
- Cleaned up unused imports and commented code.
- Adjusted logic in Home activity to reflect changes in route handling.

Commit:
feat: Improve client and itinerary handling

- Added null checks for company names to prevent crashes.
- Updated comments for clarity on loading routes.
- Enhanced itinerary display with better formatting.
- Introduced logging for debugging list retrievals.
- Adjusted layout heights for spinners in XML files.
- Set default values in login fields for testing purposes.
- Created a new spinner item layout for better UI consistency.

Commit:
chore: clean up unused imports and XML attributes

Removed unnecessary imports from several Java files to streamline the codebase. Also, cleaned up an XML layout by removing redundant namespace declarations.

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java:
```diff
@@ -56,6 +56,9 @@ protected void onCreate(Bundle savedInstanceState) {
             emailInput.setText(intent.getStringExtra(EMAIL_KEY));
         }
+
+        emailInput.setText("t@t.tt");
```
Remove that

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/entity/Client.java:
```diff
@@ -118,7 +120,7 @@ public void setId(int id) {
 
     public String getCompanyName() {
-        return companyName;
+        return companyName != null ? companyName : "";
```
Why do we check null here ?

I think All the constructor of Client are initialize `companyName`

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/entity/Client.java:
```diff
@@ -296,6 +300,23 @@ public static Client createClientFromShortJson(JSONObject clientJson) throws JSO
+    public static String getClientsDisplay(ArrayList<Client> clients) {
+        StringBuilder clientNames = new StringBuilder();
+        for (int i = 0; i < clients.size(); i++) {
+            clientNames.append(i + 1).append(". ").append(clients.get(i).toShortString());
```
I think, to build this String you can use a `String.format()` instead of appending 3 times.
It would be easier to read the string built as a developer

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/FragmentItineraries.java:
```diff
@@ -150,4 +154,22 @@ public interface FragmentItineraryActions {
+        Log.d(TAG, "getListItineraries: Get list of itineraries");
+        ArrayList<Itinerary> listItineraries = new ArrayList<>();
+        for (int i = 0; listItinerariesView.getAdapter().getCount() > i; i++) {
+            Log.d(TAG, "getListItineraries: Itinerary: " + listItinerariesView.getAdapter().getItem(i));
```
Remove this Log

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/entity/Itinerary.java:
```diff
@@ -1,4 +1,6 @@
-package fr.iut_rodez.pathpilot_android_client.home.itinerary;
+package fr.iut_rodez.pathpilot_android_client.home.itinerary.entity;
+
+import static fr.iut_rodez.pathpilot_android_client.home.clients.entity.Client.getClientsDisplay;
```
Don't import this function with static.
It can be confusing as we assume the function called is defined in `Itinerary` instead of `Client`.

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/AddRoute.java:
```diff
@@ -0,0 +1,136 @@
+
+        listItineraries.addAll(itineraries);
+
+        Log.d(TAG, "list itineraries" + listItineraries);
```
Remove this Log ?

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/FragmentRoutes.java:
```diff
@@ -0,0 +1,157 @@
+//            RouteService.deleteRoute(homeActivity, routeSelected, listRoutesView);
+//        } else {
+//            Log.e(TAG, "onContextItemSelected: Unknown option selected");
+//        }
```
Remove this commented code

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/InfoRoute.java:
```diff
@@ -0,0 +1,166 @@
+        // Convert route stops to timeline items
+        for (RouteClient routeClient : clients) {
+            timelineItems.add(new TimelineItem(routeClient));
+            Log.d(TAG, "setUpTimelineClients: Client: " + routeClient);
```
Remove this Log

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/InfoRoute.java:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    private void resumeRoute() {
```
Add comment

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/TimelineItem.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Class to represent a timeline item
+ */
+public class TimelineItem {
```
This class can be converted to a record.
Something like that
```java
public record TimelineItem(String clientName, String clientType, String address, ClientState state) {
    /**
     * Constructor for a timeline item
     *
     * @param routeClient RouteClient to create the timeline item from
     */
    public TimelineItem(RouteClient routeClient) {
        this.clientName = routeClient.getClient().getCompanyName();
        this.clientType = routeClient.getClient().getClientCategory();
        this.address = routeClient.getClient().getAddressDisplayName();
        this.state = routeClient.getState();
    }
```

Odonata971 replied:
ok its done

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/entity/Route.java:
```diff
@@ -0,0 +1,325 @@
+
+            routeBeginDate.setText(MessageFormat.format("{0}{1}", context.getString(R.string.route_begin_date), route.getDateDisplayName()));
+
+            //TODO change this according the real route state in db
```
Does this TODO done ?

Odonata971 replied:
No, there's no state in the route for the moment in the db

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/routes/service/RouteService.java:
```diff
@@ -0,0 +1,148 @@
+public class RouteService implements IRouteService {
+
+    public static final String API_BASE_URL = BuildConfig.API_BASE_URL + "routes";
+    private static final String TAG = ClientService.class.getSimpleName();
```
These two constants can be declared in the interface.

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/player/PlayerItinerary.java:
```diff
@@ -109,7 +108,11 @@ private void initialiseMap() {
+        ArrayList<Client> clients = new ArrayList<>();
+        for (int i = 0; i < route.getClients().size(); i++) {
+            clients.add(route.getClients().get(i).getClient());
+        }
```
This for loop can be improved with a stream or a for each loop.

Now at each loop, we get the complete list of client.

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/player/PlayerItinerary.java:
```diff
@@ -200,12 +203,12 @@ private void setExpectedClientMarker(ArrayList<Client> expectedClients, @NonNull
+        client.getClient().setAddressDisplayName(this);
+        clientName.setText(client.getClient().getCompanyName());
+        clientAddress.setText(client.getClient().getAddressDisplayName());
+        clientDistance.setText(getString(R.string.distance_in_km, distanceToClient(client.getClient())));
```
Extract `client.getClient()` as a new variable

Commit:
fix: change review comments

- Simplified company name retrieval logic.
- Updated client display formatting using MessageFormat.
- Streamlined client list processing with forEach and streams.
- Removed unnecessary logging statements to clean up the code.
- Enhanced timeline item representation by converting it to a record.
- Improved visit status determination with a switch expression.
- Cleaned up context menu handling in routes fragment.

Commit:
feat: Add JSON exception handling in client creation

- Updated createAccount method to handle JSONException.
- Modified addClient and sendInformationToCreateClient methods to throw JSONException.
- Simplified toJson methods in Client and Itinerary classes by removing try-catch blocks.
- Changed getClientsDisplay method to use List instead of ArrayList for better flexibility.
- Enhanced spinner item text display logic in itinerary management.

Francois389 reviewed (approved):
🗿👻🐼👀🍔🛻
